scripts: Add Renode emulation support.

mateusz-holenko opened the pull request:
This adds support for running LiteX in Renode emulation framework.

mithro commented:
The conda issue should be fixed by https://github.com/timvideos/litex-buildenv/pull/129

futaris commented:
Not sure that litex_picorv32 external interrupts are hooked up correctly, particularly for TARGET=base

```
export CPU=picorv32 PLATFORM=arty TARGET=net
source ./scripts/enter-env.sh
./scripts/build/renode.sh
```
works

```
export CPU=picorv32 PLATFORM=arty TARGET=base
source ./scripts/enter-env.sh
./scripts/build/renode.sh
```
hangs

It also hangs if you remove the eth node from /opt/renode/platforms/cpus/litex_picorv32.repl ...

mithro commented:
You'll need to rebase to make Travis work.

mateusz-holenko commented:
> Not sure that litex_picorv32 external interrupts are hooked up correctly, particularly for TARGET=base
> 
> ```
> export CPU=picorv32 PLATFORM=arty TARGET=net
> source ./scripts/enter-env.sh
> ./scripts/build/renode.sh
> ```
> works

It works only partially as you can't write on UART.

> 
> ```
> export CPU=picorv32 PLATFORM=arty TARGET=base
> source ./scripts/enter-env.sh
> ./scripts/build/renode.sh
> ```
> hangs
> 
> It also hangs if you remove the eth node from /opt/renode/platforms/cpus/litex_picorv32.repl ...

I verified the irq mapping in `platforms/cpus/litex_picorv32.repl` and you are right, it's not compatible with the defaults generated by LiteX, but...

I believe that there is a problem with overlapping interrupts in LiteX + PicoRV32.
LiteX uses IRQ1 for timer, but in PicoRV32 external IRQ1 is shared with internal `ecall/ebreak/illegal instruction`.
This caused problems when I was working on Zephyr, so I modified LiteX to use different set of IRQs for PicoRV32.
The problem was I hadn't created any PR with the change in LiteX, but published the non-standard mapping in Renode.

To sort it out, I've just crated a PR in [LiteX](https://github.com/cliffordwolf/picorv32#custom-instructions-for-irq-handling).

There is an issues though, BuildEnv provides its own mapping of peripheral's IRQs and for modified LiteX there is a conflict in `targets/arty/net.py` (and possibly other files as well).
I fixed it locally by changing `ethmac` IRQ to 6 (and this is the value in Renode's `repl` file), but I expect more fixes are needed to make it right.

mithro commented:
I rebased this onto master to get travis to run.

futaris commented:
https://github.com/enjoy-digital/litex/pull/174 for the issue mentioned above.

mithro commented on scripts/build-renode.sh:
```diff
@@ -0,0 +1,170 @@
+add_to_script "showAnalyzer sysbus.uart"
+
+# BIOS
+if grep -q 'ROM_BASE 0x00000000' $TARGET_BUILD_DIR/software/include/generated/mem.h; then
```
The ROM_BASE isn't always at memory address zero.

mithro replied:
https://github.com/timvideos/HDMI2USB-firmware-prebuilt/blob/52959d0bfe90b7536da7e516120c586e6ea3aafc/archive/master/v0.0.4-455-gde59df1/ice40_up5k_b_evn/base/lm32.minimal/software/include/generated/mem.h#L7

mateusz-holenko replied:
You are right, I fixed it and now the actual value is read from `generated/mem.h`.

I took this from `scripts/build-qemu.sh` though, where this value is hard-coded.

mithro commented on scripts/build-renode.sh:
```diff
@@ -0,0 +1,170 @@
+	echo "Platform has BIOS ROM, adding BIOS"
+
+	add_to_script "sysbus LoadBinary @$TARGET_BUILD_DIR/software/bios/bios.bin 0x0"
+	add_to_script "cpu PC 0x0"
```
The reset vector isn't always zero.

mithro replied:
https://github.com/timvideos/HDMI2USB-firmware-prebuilt/blob/52959d0bfe90b7536da7e516120c586e6ea3aafc/archive/master/v0.0.4-455-gde59df1/ice40_up5k_b_evn/base/lm32.minimal/software/include/generated/csr.h#L322-L325

mateusz-holenko replied:
In the fixed version, I assumed the reset vector value to be the same as `ROM_BASE`.

pgielda commented:
I did some tests.
I would suggest adding two things:
1) Silencing SDRAM_CONTROLLER reads/writes so that the console is not flooded.
2) add "showAnalyzer uart Renode.Analyzers.LoggingUartAnalyzer" so that console output is printed with the rest of the log

pgielda commented:
We should also tag this peripheral:

02:17:37.1529 [WARNING] sysbus: [cpu: 0xC50] WriteDoubleWord to non existing peripheral at 0xE0005804, value 0x1.
02:17:37.1529 [WARNING] sysbus: [cpu: 0xC54] WriteDoubleWord to non existing peripheral at 0xE0005808, value 0x1.
02:17:37.1530 [WARNING] sysbus: [cpu: 0xC58] WriteDoubleWord to non existing peripheral at 0xE0005810, value 0x1.
02:17:37.1531 [WARNING] sysbus: [cpu: 0xC60] WriteDoubleWord to non existing peripheral at 0xE0005804, value 0x2.
02:17:37.1532 [WARNING] sysbus: [cpu: 0xC64] WriteDoubleWord to non existing peripheral at 0xE0005808, value 0x1.
02:17:37.1532 [WARNING] sysbus: [cpu: 0xC6C] WriteDoubleWord to non existing peripheral at 0xE0005810, value 0x1.

And most probably silence this too.

mateusz-holenko commented:
In the newest version:

- if Renode is not found in the system, version 1.7 is downloaded from github and copied to `build/renode` directory

- platform definition file is generated dynamically from `csr.h` and `mem.h` files

- access to some peripherals is silenced (so it does not flood the console)

- uart output is logged

mithro commented on scripts/build-renode.sh:
```diff
@@ -0,0 +1,265 @@
+	fi
+fi
+
+script_file=$(mktemp /tmp/litex_buildenv.resc.XXXXXX)
```
Can we move the code generating the configuration to a Python script? It would be much nicer than this then?

mateusz-holenko replied:
I also moved generation of resc commands to the same python script.

mithro commented on scripts/build-renode.sh:
```diff
@@ -0,0 +1,265 @@
+
+RENODE_BIN="renode"
+command -v $RENODE_BIN 2>&1 1>/dev/null
+if [ $? -ne 0 ]; then
```
Please add a comment here. I'm assuming this is downloading a prebuilt renode release? On what platforms will this run?

mateusz-holenko replied:
You are right, this is a pre-built Renode compiled for 64-bit Linux. It should run on any distro assuming dependencies are present in the system (list of runtime dependecies is described [here](https://github.com/renode/renode/blob/master/README.rst#installation).

Other option is to install package (that would take care of dependencies automatically). The downside is that:

- it's installed globally in the system, polluting it a little bit,
- packages are distro-specific and we currently support: debian/ubuntu, fedora and arch so we would have to detect the distribution and install appropriate package; the question what about other distributions?

mithro commented on scripts/build-renode.sh:
```diff
@@ -0,0 +1,265 @@
+set -x
+set -e
+
+RENODE_BIN="renode"
```
Allow people to override the location of renode binary to use.

mithro commented on scripts/build-renode.sh:
```diff
@@ -0,0 +1,162 @@
+	make firmware
+fi
+
+function parse_generated_header {
```
This is trying to get the IP address right? I would suggest we use the following name...
```suggestion
function get_ethernet_ip_address {
```

This is probably useful in the qemu emulation too. We should do that there too?

mateusz-holenko replied:
Before moving to a python script this function was used to extract different values defined in `mem.h`/`csr.h` files. Right now it's used in just one context, so it lost its universality - refactoring and renaming it seems like a good idea.

The whole tap-related fragment of script is a copy-paste from `build-qemu.sh`, so the two scripts shares a lot more code. I will extract it to some base file and source in both `build-renode.sh` and `build-qemu.sh`.

mateusz-holenko replied:
After extracting the common function to `build-common.sh` it turns out that `parse_generated_header` function might be used in `build-qemu` in several places (in different context), so I kept it's name.

mithro commented on scripts/build-renode.sh:
```diff
@@ -0,0 +1,162 @@
+fi
+
+script_file=$(mktemp /tmp/litex_buildenv.resc.XXXXXX)
+platform_file=$(mktemp /tmp/litex_buildenv.repl.XXXXXX)
```
Lets put these files in the build directory under `build/<build directory>/renode`?

mithro replied:
We then don't need to clean them up and people can launch renode manually with the configurations.

mateusz-holenko replied:
This is a good idea.

mateusz-holenko replied:
It's done.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+
```
Please add top level doc string.

mateusz-holenko replied:
Done.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+import zlib
+import argparse
+
+def generate_ethmac(data):
```
Please add docstring and [maybe type hints](https://docs.python.org/3/library/typing.html).

mateusz-holenko replied:
I added docstrings to all functions.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+    buffer_size = get_entry_value(data, 'memory_region', 'ethmac', 1)
+    interrupt = get_entry_value(data, 'constant', 'ethmac_interrupt')
+
+    result = [
```
Use a triple quoted string for long output like this.

mateusz-holenko replied:
This is reworked to return a single, multiline string.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+    base = 0x20228000
+    return [
+        'boot_flash: Memory.ArrayMemory @ sysbus {}'.format(base),
+        '    size: 0x00100000'
```
The size / base comes from `mem.h` in `build/XXXX/software/include/generated/`
```c
#define SPIFLASH_BASE 0xa0000000
#define SPIFLASH_SIZE 0x01000000
```

mithro replied:
This is actually in the CSV file in `build/XXX/test/csr.csv` too;
```csv
memory_region,rom,0x00000000,32768,
memory_region,sram,0x10000000,32768,
memory_region,spiflash,0xa0000000,16777216,
memory_region,main_ram,0x40000000,268435456,
memory_region,ethmac,0xb0000000,8192,
```

mateusz-holenko replied:
I was confused by the fact that in `include/generated/mem.h` `FLASH_BOOT_ADDRESS` is defined as `0x20228000`, but `spiflash` is located at `0xa0000000`.

Now I see in `targets/arty/base.py` that this address is a default shadow for `0x20000000`.
For Arty platform, the flash boot address is `spi_flash` base + gateware size (0x220000) + bios size (0x8000), so it all adds up.

It's still a bit problematic, as needed parameters (information about shadow address, gateware size or bios size) are not included in `csr.csv` so I don't see a way of calculating the final value based solely on this input.

mateusz-holenko replied:
I looked a little in LiteX sources and it looks that the best option is to emit more information (i.e., `shadow_base` and `flash_boot_offset`) to `csr.csv` file.

As it will require a separate PR to LiteX that would block this one, I decided to temporarily remove the functionality of booting from flash (still leaving a netboot option available). I will add it with a separate PR, once the necessary code is in LiteX.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+
+def generate_repl(data):
+    return \
+        generate_cpu(data) + \
```
Don't use line continuations like this;
```python

def generate_repl(data):
 output = []
 output += generate_cpu(data)
 output += xxx
 return "".join(output)
```

mateusz-holenko replied:
This fragment is totally reworked.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+        result.append('    -> cpu@{}'.format(interrupt))
+    return result
+
+def generate_peripheral(data, name, model, properties=None):
```
What if the peripheral doesn't exist?

mateusz-holenko replied:
My idea was to return an empty array that would concatenate nicely, but now the generating code is totally reworked, so this is no more valid.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+    return \
+        generate_cpu(data) + \
+        generate_memory_region(data, 'rom') + \
+        generate_memory_region(data, 'sram') + \
```
You should loop over the memory regions define in either `mem.h` or `csr.csv`;
```python

data = """
memory_region,rom,0x00000000,32768,
memory_region,sram,0x10000000,32768,
memory_region,spiflash,0xa0000000,16777216,
memory_region,main_ram,0x40000000,268435456,
memory_region,ethmac,0xb0000000,8192,
"""
for line in csv.reader(data):
  if line[0] != "memory_region":
    _, memory_name, memory_start_adress, memory_size = *line
   # Some peripherals (like Ethernet) define a memory region, as peripherals will already be defined and taken ownership of the region we skip creating it here.
   if model.already_has_memory(memory_name):
      print("Skipping {} as already claimed by peripheral {}".format(memory_name, model.already_has_memory(memory_name))
      continue
    generate_memory_region(memory_name, memory_start_address, memory_size)
```

mateusz-holenko replied:
I created a separate `parse_csv` function that does the job.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+            'sysbus WriteDoubleWord {} {}'.format(hex(flash_boot_address), hex(len(firmware_data))),
+            'sysbus WriteDoubleWord {} {}'.format(hex(flash_boot_address + 4), hex(crc32)),
+            'sysbus LoadBinary @{} {}'.format(firmware_binary, hex(flash_boot_address + 8))
+        ])
```
Can you add a comment about what this is doing and why?

mateusz-holenko replied:
This fragment of code is temporarily removed. See comments above.

mithro commented on scripts/generate-renode-scripts.py:
```diff
@@ -0,0 +1,168 @@
+    return None
+
+def generate_repl(data):
+    return \
```
You want code that does something like this;

```python
#! /usr/bin/env python3

data = """
csr_base,cas,0xe0006800,,
csr_base,ctrl,0xe0000000,,
csr_base,ddrphy,0xe0005800,,
csr_base,ethmac,0xe0007800,,
csr_base,ethphy,0xe0007000,,
csr_base,info,0xe0006000,,
csr_base,sdram,0xe0004000,,
csr_base,spiflash,0xe0005000,,
csr_base,timer0,0xe0002800,,
csr_base,uart,0xe0001800,,
csr_base,uart_phy,0xe0001000,,
csr_base,identifier_mem,0xe0002000,,
csr_register,cas_leds_out,0xe0006800,1,rw
csr_register,cas_switches_in,0xe0006804,1,ro
csr_register,cas_buttons_ev_status,0xe0006808,1,rw
csr_register,cas_buttons_ev_pending,0xe000680c,1,rw
csr_register,cas_buttons_ev_enable,0xe0006810,1,rw
constant,remoteip1,192,,
constant,remoteip2,168,,
constant,remoteip3,100,,
constant,remoteip4,100,,
constant,cas_leds_count,4,,
constant,cas_switches_count,4,,
constant,cas_buttons_count,4,,
constant,ethmac_rx_slots,2,,
constant,ethmac_tx_slots,2,,
"""

# Build a list of peripherals and collect all CSR registers + CSR constants associated with the peripherals
global_constants = {}
peripherals = {}
for line in csv.reader(data):
    if line[0] == "csr_base":
        _, peripheral_name, peripheral_register_base = line
        assert peripheral_name not in peripherals
        registers = {}
        constants = {}
        peripherals[peripheral_name] = (peripheral_register_base, registers, constants)

    elif line[0] == "csr_register":
        _, register_name, register_address, register_width, register_mode = line
        for peripheral_name in peripherals.keys():
            if not register_name.startswith(peripheral_name):
                continue
            peripheral_register_base, peripheral_registers, _ = peripherals[peripheral_name]

            # Strip the peripheral prefix
            register_short_name = constant_name[len(peripheral_name)+1:]

            register_offset = register_address - peripheral_register_base
            assert register_short_name not in peripheral_registers
            peripheral_registers[register_name] = (register_offset, register_width, register_mode)
            break
        else:
            # CSR Registers must be associated with a peripheral
            assert False, "CSR Register {} not associated with any peripheral {}".format(register_name, peripherals.keys())

    elif line[0] == "constant":
        _, constant_name, constant_value = line
        for peripheral_name in peripherals.keys():
            if not constant_name.startswith(peripheral_name):
                continue

            # Strip the peripheral prefix
            constant_short_name = constant_name[len(peripheral_name)+1:]

            peripheral_register_base, _, peripheral_constants = peripherals[peripheral_name]
            assert constant_short_name not in peripheral_constants
            peripheral_constants[constant_short_name] = constant_value

            break
        else:
            global_constants[constant_name] = constant_value

    else:
        assert False, "Unknown line '{}'".format(line)

import pprint
pprint.pprint(global_constants)
pprint.pprint(peripherals)

for peripheral_name, (peripheral_register_base, peripheral_registers, peripheral_constants) in peripherals.values():
    peripheral_handler = detect_peripheral_handler(peripheral_registers)
    if peripheral_handler is None:
        peripheral_hanlder = mock_handler(peripheral_register_base, peripheral_registers)

    peripheral_handler(config_file, peripheral_name, peripheral_register_base, peripheral_registers, peripheral_constants)
```

mateusz-holenko replied:
I created a separate parse_csv function that does the job. I skipped parsing registers as it's not needed for Renode.

mateusz-holenko commented:
> > Not sure that litex_picorv32 external interrupts are hooked up correctly, particularly for TARGET=base
> > ```
> > export CPU=picorv32 PLATFORM=arty TARGET=net
> > source ./scripts/enter-env.sh
> > ./scripts/build/renode.sh
> > ```
> > 
> > 
> > works
> 
> It works only partially as you can't write on UART.
> 
> > ```
> > export CPU=picorv32 PLATFORM=arty TARGET=base
> > source ./scripts/enter-env.sh
> > ./scripts/build/renode.sh
> > ```
> > 
> > 
> > hangs
> > It also hangs if you remove the eth node from /opt/renode/platforms/cpus/litex_picorv32.repl ...
> 
> I verified the irq mapping in `platforms/cpus/litex_picorv32.repl` and you are right, it's not compatible with the defaults generated by LiteX, but...
> 
> I believe that there is a problem with overlapping interrupts in LiteX + PicoRV32.
> LiteX uses IRQ1 for timer, but in PicoRV32 external IRQ1 is shared with internal `ecall/ebreak/illegal instruction`.
> This caused problems when I was working on Zephyr, so I modified LiteX to use different set of IRQs for PicoRV32.
> The problem was I hadn't created any PR with the change in LiteX, but published the non-standard mapping in Renode.
> 
> To sort it out, I've just crated a PR in [LiteX](https://github.com/cliffordwolf/picorv32#custom-instructions-for-irq-handling).
> 
> There is an issue though, BuildEnv provides its own mapping of peripheral's IRQs and for modified LiteX there is a conflict in `targets/arty/net.py` (and possibly other files as well).
> I fixed it locally by changing `ethmac` IRQ to 6 (and this is the value in Renode's `repl` file), but I expect more fixes are needed to make it right.

The ideas from my PR has been merged to LiteX master with [this commit](https://github.com/enjoy-digital/litex/commit/47dc87584f2bc29e837b4319f516ad3369473fd9).

@mithro I see that buildenv LIteX submodule is a bit behind master. Do you knwo if switching to a newer version would be a problem?

mithro commented:
@mateusz-holenko To update the submodules it should be as simple as going into `./third_party` directory and running `/merge-upstream.sh`.

mateusz-holenko commented:
@mithro Did you have time to look at the current version of the script?

mateusz-holenko commented:
I rebased the code onto the current master and changed Renode binary source to nightly builds.

mateusz-holenko commented:
@mithro Travis failed, but it does not seem to be related to my changes. How can I force it to run once again?
